xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-api/src/main/java/org/xwiki/notifications/internal/page/DefaultPageNotificationDisplayer.java

```diff
@@ -0,0 +1,153 @@
+
+/**
+ * Implement a {@link NotificationDisplayer} for the event type
+ * {@link org.xwiki.notifications.page.PageNotificationEvent}.
+ *
+ * @version $Id$
+ * @since 9.4RC1
+ */
+@Component
+@Singleton
+@Named(DefaultPageNotificationDisplayer.EVENT_TYPE)
+public class DefaultPageNotificationDisplayer implements NotificationDisplayer
+{
+    /**
+     * Event handled by the displayer.
+     */
+    public static final String EVENT_TYPE = "org.xwiki.notifications.page.PageNotificationEvent";
+
+    private static final String EVENT_BINDING_NAME = "event";
+
+    @Inject
+    private PageNotificationEventDescriptorContainer pageNotificationEventDescriptorContainer;
+
+    @Inject
+    private Provider<XWikiContext> contextProvider;
+
+    @Inject
+    private TemplateManager templateManager;
+
+    @Inject
+    private VelocityManager velocityManager;
+
+    @Inject
+    private AuthorExecutor authorExecutor;
+
+    private class VelocityTemplateStringEvaluator implements Callable<String>
+    {
+        private String content;
+
+        private Provider<XWikiContext> contextProvider;
+
+        /**
+         * Constructs a {@link VelocityTemplateStringEvaluator}.
+         *
+         * @param contextProvider a reference to the XWikiContext provider component
+         * @param content the content of the template that will be evaluated
+         */
+        VelocityTemplateStringEvaluator(Provider<XWikiContext> contextProvider, String content)
+        {
+            this.contextProvider = contextProvider;
+            this.content = content;
+        }
+
+        @Override
+        public String call() throws Exception
+        {
+            return contextProvider.get().getWiki().evaluateVelocity(content, "page-notification");
+        }
+    }
+
+    @Override
+    public Block renderNotification(CompositeEvent eventNotification) throws NotificationException
+    {
+        try {
+            velocityManager.getCurrentVelocityContext().put(EVENT_BINDING_NAME, eventNotification);
+
+            PageNotificationEventDescriptor eventDescriptor =
+                    pageNotificationEventDescriptorContainer.getDescriptorByType(
+                            eventNotification.getEvents().get(0).getType());
+
+            // If we have no template defined, use the default one
+            if (eventDescriptor.getNotificationTemplate().isEmpty()) {
+                return templateManager.execute("notification/default.vm");
+            }
+
+            return new RawBlock(executeTemplate(eventDescriptor.getAuthorReference(),
+                    eventDescriptor.getNotificationTemplate()), Syntax.HTML_5_0);
+        } catch (Exception e) {
+            throw new NotificationException("Unable to render notification template.", e);
+        } finally {
+            velocityManager.getCurrentVelocityContext().remove(EVENT_BINDING_NAME);
+        }
+    }
+
+    private String executeTemplate(DocumentReference userReference, String templateContent)
```

Comment:
https://github.com/xwiki/xwiki-platform/tree/master/xwiki-platform-core/xwiki-platform-template/xwiki-platform-template-api/src/main/java/org/xwiki/template only contains API which are implemented by https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/internal/template/InternalTemplateManager.java.

> So I’m thinking about writing a new implementation of Template that would allow me to instantiate a template by directly passing the template content and the template author to the constructor.

Implementing Template is no much but sure your use case is generic enough to provide something like a StringTemplate helper somewhere in https://github.com/xwiki/xwiki-platform/tree/master/xwiki-platform-core/xwiki-platform-template/xwiki-platform-template-api/src/main/java/org/xwiki/template yes.